### ISSUE TITLE
Add custom parameter to WMSTileLayer

### ISSUE THREAD

**Issue by prl900** [thedirtyfew/dash-leaflet#90]
First of all, thank you for contributing this library, is a fantastic project. 

Some WMS accept url parameters that are not in the standard, or are only available in certain implementations. An example of this is the `time` parameter, which is often used to specify a time for a layer. The current WMSTileLayer function does not allow passing custom parameters, which unfortunately prevents making use of this extra functionality on WMS servers.

I tried to hack this limitation by injecting extra parameters in the value of an available parameter. For example:
`format="image/png&time=2020"`

Unfortunately, the special url characters (&, =) get escaped and this does not work.

It would be a nice and very useful extension for this function to be able to accept custom parameters. A **kwargs argument at the end where users can define their own parameters?

I'd be happy to contribute this piece of functionality once agreed and after receiving some guidance about what are the affected sections in the code.

Thanks

**Comment by emilhe** [comment]
Thanks! I have just made an initial implementation where you can pass extra properties to the WMSLayer via a new `extraProps` property (the `kwargs` attribute is reserved for internal Dash usage). You can try it from pypi,

pip install dash-leaflet==0.1.14rc1

Here is my test example,

```
import dash_html_components as html
import dash_leaflet as dl
from dash import Dash

wms_layer = dl.WMSTileLayer(url="https://mesonet.agron.iastate.edu/cgi-bin/wms/nexrad/n0r.cgi",
                            layers="nexrad-n0r-900913", format="image/png", transparent=True,
                            extraProps=dict(someProp="someValue"))
app = Dash()
app.layout = html.Div([
    dl.Map([dl.TileLayer(), wms_layer], center=[40, -100], zoom=4,
           style={'width': '100%', 'height': '50vh', 'margin': "auto", "display": "block"}),
])

if __name__ == '__main__':
    app.run_server()
```

I don't have access to any wms tile servers that require special keywords though, so i haven't tested the code much beyond syntax errors.

**Comment by prl900** [comment]
Wow! Thank you very much @emilhe 

Is is possible you forgot to commit this change into the new version? I'm getting this error:
```
TypeError: The `dash_leaflet.WMSTileLayer` component (version 0.1.14rc1) 
with the ID "wms_layer" received an unexpected keyword argument: `extraProps`

Allowed arguments: attribution, bounds, children, className, crossOrigin, crs, detectRetina, 
errorTileUrl, format, id, keepBuffer, layers, maxNativeZoom, maxZoom, minNativeZoom, minZoom, 
noWrap, opacity, pane, styles, subdomains, tileSize, tms, transparent, updateInterval, updateWhenIdle, 
updateWhenZooming, uppercase, url, version, zIndex, zoomOffset, zoomReverse
```

**Comment by emilhe** [comment]
Ah, sorry, it was one version number too old, it is,

`pip install dash-leaflet==0.1.14rc2`

**Comment by emilhe** [comment]
Hmm, for some reason it seems my local builds are not uploaded properly to pypi. Anyhow, I just made a build with directly on github instead,

`pip install dash-leaflet==0.1.14rc3`

**Comment by prl900** [comment]
Thank you @emilhe, you are very fast! Your previous rc2 worked perfectly for me.

**Comment by kirodh** [comment]
Hi there

Could you please provide an example how to use the extra parameters please. Have been trying but not succeeding.


Here is an example link i want to connect to:

http://......./geoserver/myworkspace/wms?service=WMS&version=1.1.0&request=GetMap&layers=myworkspace%3AmyLayer&bbox=-180.0 -90.0 180.0 90.0&width=768&height=384&srs=EPSG 4326&format=application/openlayers&time=2021-06-05&CQL_FILTER=location LIKE 'Tile_20210605T074609_yyy_rgb.tif' 

Thanks in advance

**Comment by prl900** [comment]
If you pip install the version that @emilhe created for this, you can use the example he posted:
```
wms_layer = dl.WMSTileLayer(url="https://mesonet.agron.iastate.edu/cgi-bin/wms/nexrad/n0r.cgi",
                            layers="nexrad-n0r-900913", format="image/png", transparent=True,
                            extraProps=dict(someProp="someValue"))
```

In my case, I used:
```
extraProps=dict(time="2020-05-01T00:00:00.000Z")
```

I hope it helps

**Comment by kirodh** [comment]
Hi prl900

Thanks for the quick reply earlier. I implemented your way and it still never worked (I am using the django-plotly-dash implementation). I did make a sample code out of the django-dash framework and the requests never worked until I installed the `rc2 `version, (i had to comment out the fullScreenControl in the imports file, as it threw an error), then it worked. So i installed the `rc3 `version after this and then everything seemed to work. I dont know if it was a system related issue, or probably myself not doing something right. I used the time and CQL_FILTER `extraProps` keywords. Dont know if it also contributed, but a `zIndex=999999999,` was also used.

However, when I copy the exact same code to the django-plotly-dash framework code i have, then it doesnt work. I do get an image appearing, but when filtered on time (I am using a mosaic from geoserver), i should see many tiles, (i waited for some time as it takes time to load), but nothing. In the stand alone code, I see many tiles appearing.

If you have tips, i am all ears,  will keep on working on this until i find a solution.

Thanks in advance

**Comment by prl900** [comment]
Hi @kirodh, I'd recommend trying to break down and simplify your use case to find what the problem is. WMS layers are normally composed of tiles (ie 256x256 pixels) that are stitched to cover a map area. Each of these tiles is an independent request to the server.

Here is an example of a tile request to the WMS used in the dash-leaflet [demo page](https://dash-leaflet.herokuapp.com/):
`https://mesonet.agron.iastate.edu/cgi-bin/wms/nexrad/n0r.cgi?service=WMS&request=GetMap&layers=nexrad-n0r-900913&styles=&format=image%2Fpng&transparent=true&version=1.1.1&width=256&height=256&srs=EPSG%3A3857&bbox=-8766409.899970295,5009377.085697314,-7514065.628545968,6261721.357121641`

If you click on the url you'll see the tile rendered in your browser. Also you can see the different parameters that are passed to the WMS server as arguments in the request. 

If the server that you are using accepts `time` as an argument you'll be able to test this by passing the argument manually in a similar request. All that dash-leaflet is doing here with the `extraProps` argument is to get a dictionary from the user and pass that as extra arguments straight to the server.

**Comment by emilhe** [comment]
@kirodh It is a known issue that `dash-leaflet` is not compatible with `django-plotly-dash`.

**Comment by emilhe** [comment]
The feature has now been merged and release as part of the 0.0.14 release.

https://pypi.org/project/dash-leaflet/0.1.14/
